src/dashboards/components/DashboardContainer.tsx

```diff
@@ -21,17 +21,98 @@ import {AUTOREFRESH_DEFAULT} from 'src/shared/constants'
 // Types
 import {AppState, ResourceType, AutoRefreshStatus} from 'src/types'
 
+// Notifications
+import {dashboardAutoRefreshTimeoutSuccess} from 'src/shared/copy/notifications'
+import {notify} from 'src/shared/actions/notifications'
+
+// History
+import {useRouteMatch} from 'react-router-dom'
+
 const {Active} = AutoRefreshStatus
 
 type ReduxProps = ConnectedProps<typeof connector>
 type Props = ReduxProps
 
+let timer
 const DashboardContainer: FC<Props> = ({autoRefresh, dashboard}) => {
   const dispatch = useDispatch()
+  const isEditing = useRouteMatch(
+    '/orgs/:orgID/dashboards/:dashboardID/cells/:cellID/edit'
+  )
+
+  const registerListeners = useCallback(() => {
+    if (timer) {
+      registerStopListeners()
+    }
+    timer = setTimeout(() => {
+      dispatch(resetDashboardAutoRefresh(dashboard))
+      dispatch(notify(dashboardAutoRefreshTimeoutSuccess()))
+      registerStopListeners()
+    }, autoRefresh.inactivityTimeout)
+
+    window.addEventListener('load', registerListeners)
+    document.addEventListener('mousemove', registerListeners)
+    document.addEventListener('keypress', registerListeners)
+  }, [dashboard, autoRefresh.inactivityTimeout])
+
+  const visChangeHandler = () => {
+    if (
+      document.visibilityState === 'hidden' &&
+      autoRefresh.status === AutoRefreshStatus.Active
+    ) {
+      registerListeners()
+    } else {
+      registerStopListeners()
+    }
+  }
+  const registerStopListeners = useCallback(() => {
+    // Stop all existing timers and deregister everythang
+    if (timer) {
+      clearTimeout(timer)
+      timer = null
+    }
+
+    window.removeEventListener('load', registerListeners)
+    document.removeEventListener('mousemove', registerListeners)
+    document.removeEventListener('keypress', registerListeners)
+  }, [registerListeners])
+
+  document.addEventListener('visibilitychange', visChangeHandler)
+
+  useEffect(() => {
+    if (
+      autoRefresh?.status &&
+      autoRefresh.status === AutoRefreshStatus.Active &&
+      autoRefresh.inactivityTimeout > 0
+    ) {
+      registerListeners()
+    } else {
+      registerStopListeners()
+    }
+    return () => {
+      registerStopListeners()
+      document.removeEventListener('visibilitychange', visChangeHandler)
```

Comment:
if you're going to be referencing the event listener like this, you should handle it's registration inside the useEffect as well. how it's set up now, it's going to register a new event listener on every render, then deregister it _sometimes_, which will lead to a slow memory leak at best, and calling a function a billion times at worst